executor/compiler.go

```diff
@@ -46,17 +49,168 @@ func (c *Compiler) Compile(goCtx goctx.Context, stmtNode ast.StmtNode) (*ExecStm
 		return nil, errors.Trace(err)
 	}
 
+	CountStmtNode(stmtNode, c.Ctx.GetSessionVars().InRestrictedSQL)
+	isExpensive := logExpensiveQuery(stmtNode, finalPlan)
+
 	return &ExecStmt{
 		InfoSchema: infoSchema,
 		Plan:       finalPlan,
-		Expensive:  stmtCount(stmtNode, finalPlan, c.Ctx.GetSessionVars().InRestrictedSQL),
+		Expensive:  isExpensive,
 		Cacheable:  plan.Cacheable(stmtNode),
 		Text:       stmtNode.Text(),
 		StmtNode:   stmtNode,
 		Ctx:        c.Ctx,
 	}, nil
 }
 
+func logExpensiveQuery(stmtNode ast.StmtNode, finalPlan plan.Plan) (expensive bool) {
+	expensive = isExpensiveQuery(finalPlan)
+	if !expensive {
+		return
+	}
+
+	const logSQLLen = 1024
+	sql := stmtNode.Text()
+	if len(sql) > logSQLLen {
+		sql = fmt.Sprintf("%s len(%d)", sql[:logSQLLen], len(sql))
+	}
+	log.Warnf("[EXPENSIVE_QUERY] %s", sql)
+	return
+}
+
+func isExpensiveQuery(p plan.Plan) bool {
+	switch x := p.(type) {
+	case plan.PhysicalPlan:
+		return isPhysicalPlanExpensive(x)
+	case *plan.Execute:
+		return isExpensiveQuery(x.Plan)
+	case *plan.Insert:
+		if x.SelectPlan != nil {
+			return isPhysicalPlanExpensive(x.SelectPlan)
+		}
+	case *plan.Delete:
+		if x.SelectPlan != nil {
+			return isPhysicalPlanExpensive(x.SelectPlan)
+		}
+	case *plan.Update:
+		if x.SelectPlan != nil {
+			return isPhysicalPlanExpensive(x.SelectPlan)
+		}
+	}
+	return false
+}
+
+func isPhysicalPlanExpensive(p plan.PhysicalPlan) bool {
+	const expensiveRowThreshold = 10000
+	if p.StatsInfo().Count() > expensiveRowThreshold {
+		return true
+	}
+
+	for _, child := range p.Children() {
+		if isPhysicalPlanExpensive(child) {
+			return true
+		}
+	}
+
+	switch x := p.(type) {
+	case *plan.PhysicalTopN:
+		return true
+	case *plan.PhysicalSort:
+		return true
+	case *plan.NominalSort:
```

Comment:
Why is a nominal sort expensive?

Reply:
"NominalSort" should not appear in the final physical plan, removed.